File: Serenity.Script.UI/FilterPanel/FilteringTypes/EditorFiltering.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using jQueryApi;

namespace Serenity
{
    [Imported(ObeysTypeSystem = true)]
    public class EditorFiltering : BaseEditorFiltering
    {
        public EditorFiltering()
            : base(typeof(Widget))
        {
        }

        [Option, IntrinsicProperty]
        public string EditorType { get; set; }
        [Option, IntrinsicProperty]
        public bool UseRelative { get; set; }
        [Option, IntrinsicProperty]
        public bool UseLike { get; set; }

        public override List<FilterOperator> GetOperators()
        {
            var list = new List<FilterOperator>();
            list.Add(FilterOperators.EQ);
            list.Add(FilterOperators.NE);

            if (UseRelative)
            {
                list.Add(FilterOperators.LT);
                list.Add(FilterOperators.LE);
                list.Add(FilterOperators.GT);
                list.Add(FilterOperators.GE);
            }

            if (UseLike)
            {
                list.Add(FilterOperators.Contains);
                list.Add(FilterOperators.StartsWith);
            }

            AppendNullableOperators(list);

            return list;
        }

        protected override bool UseEditor()
        {
            return Operator.Key == FilterOperators.EQ ||
                Operator.Key == FilterOperators.NE ||
                (UseRelative &&
                 (Operator.Key == FilterOperators.LT ||
                  Operator.Key == FilterOperators.LE ||
                  Operator.Key == FilterOperators.GT ||
                  Operator.Key == FilterOperators.GE));
        }

        protected override object GetEditorOptions()
        {
            var opt = base.GetEditorOptions();

            if (UseEditor() && EditorType == (Field.EditorType ?? "String"))
                opt = jQuery.Extend(opt, Field.EditorParams);

            return opt;
        }

        public override void CreateEditor()
        {
            if (UseEditor())
            {
                var editorType = EditorTypeRegistry.Get(EditorType);
                editor = Widget.CreateOfType(editorType, e => e.AppendTo(Container), GetEditorOptions());
                return;
            }

            base.CreateEditor();
        }

        protected override bool UseIdField()
        {
            return UseEditor();
        }

        protected override string GetEditorText()
        {
            return base.GetEditorText();
        }

        public override void InitQuickFilter(QuickFilter<Widget, object> filter)
        {
            base.InitQuickFilter(filter);

            filter.Type = EditorTypeRegistry.Get(EditorType);
        }
    }
}

File: packages/corelib/src/ui/filtering/editorfiltering.ts
import { nsSerenity } from "../../base";
import { EditorTypeRegistry } from "../../types/editortyperegistry";
import { QuickFilter } from "../datagrid/quickfilter";
import { Widget } from "../widgets/widget";
import { BaseEditorFiltering } from "./baseeditorfiltering";
import { FilterOperator, FilterOperators } from "./filteroperator";

/**
 * A filtering handler that uses an editor type resolved from the editor type registry.
 */
export class EditorFiltering extends BaseEditorFiltering<Widget<any>> {
    static override[Symbol.typeInfo] = this.registerClass(nsSerenity);

    /**
     * Creates an editor filtering handler.
     * @param props - Options including the editor type and operator flags.
     */
    constructor(public readonly props: { editorType?: string, useRelative?: boolean, useLike?: boolean } = {}) {
        super(Widget);
        this.props ??= {};
    }

    /** The editor type key. */
    get editorType() { return this.props.editorType }
    /** Sets the editor type key. */
    set editorType(value) { this.props.editorType = value }

    /** Whether relative comparison operators are used. */
    get useRelative() { return this.props.useRelative }
    /** Sets whether relative comparison operators are used. */
    set useRelative(value) { this.props.useRelative = value }

    /** Whether like operators (contains, startsWith) are used. */
    get useLike() { return this.props.useLike }
    /** Sets whether like operators are used. */
    set useLike(value) { this.props.useLike = value }

    /**
     * Returns the operators supported by this filtering handler.
     * @returns The operators.
     */
    getOperators(): FilterOperator[] {
        var list = [];

        list.push({ key: FilterOperators.EQ });
        list.push({ key: FilterOperators.NE });

        if (this.useRelative) {
            list.push({ key: FilterOperators.LT });
            list.push({ key: FilterOperators.LE });
            list.push({ key: FilterOperators.GT });
            list.push({ key: FilterOperators.GE });
        }

        if (this.useLike) {
            list.push({ key: FilterOperators.contains });
            list.push({ key: FilterOperators.startsWith });
        }

        this.appendNullableOperators(list);

        return list;
    }

    /**
     * Whether the current operator uses an editor.
     * @returns True when an editor is used.
     */
    protected override useEditor() {
        var op = this.get_operator().key;

        return op === FilterOperators.EQ ||
            op === FilterOperators.NE ||
            (this.useRelative && (
                op === FilterOperators.LT ||
                op === FilterOperators.LE ||
                op === FilterOperators.GT ||
                op === FilterOperators.GE));
    }

    /**
     * Creates the editor for the current operator.
     */
    override createEditor() {
        if (this.useEditor()) {
            var editorType = EditorTypeRegistry.get(this.editorType ?? 'String') as typeof Widget<{}>;

            this.editor = new editorType({
                element: el => this.get_container().append(el),
                ...this.getEditorOptions()
            }).init?.();

            return;
        }

        super.createEditor();
    }

    /**
     * Whether to use the id field for the criteria.
     * @returns True when an editor is used.
     */
    protected override useIdField(): boolean {
        return this.useEditor();
    }

    /**
     * Initializes a quick filter using the editor type.
     * @param filter - The quick filter to initialize.
     */
    override initQuickFilter(filter: QuickFilter<Widget<any>, any>) {
        super.initQuickFilter(filter);

        filter.type = EditorTypeRegistry.get(this.editorType ?? 'String');
    }
}
